### ISSUE TITLE
Order of the elements in the queue changes unexpectedly

### ISSUE THREAD

**Issue by jmlipman** [fepegar/torchio#991]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug summary

I noticed that the order of the elements in the queue changes when I run a validation loop inside the training loop. I guess that the order in which the training set is drawn should not change since the validation loop iterates over a different set of images (the validation set).

### Code for reproduction

```python
import torchio as tio                                                           
import numpy as np                                                              
from torch.utils.data.dataloader import DataLoader                              
import torch, random                                                            
                                                                                
seed = 42                                                                       
torch.manual_seed(seed)                                                         
torch.cuda.manual_seed(seed)                                                    
np.random.seed(seed)                                                            
random.seed(seed)                                                               
                                                                                
subjects = []                                                                   
for sub_id in range(1, 11):                                                     
    params = {                                                                  
        "im": tio.ScalarImage(tensor=np.random.random((1, 320, 320, 10))),         
        "num_samples": 8,                                                       
        "info": str(sub_id)                                                     
    }                                                                           
    subjects.append(tio.Subject(**params))                                      
                                                                                
# Shameless copypaste                                                           
subjects_val = []                                                               
for sub_id in range(12, 20):                                                    
    params = {                                                                  
        "im": tio.ScalarImage(tensor=np.random.random((1, 320, 320, 10))),         
        "num_samples": 8,                                                       
        "info": str(sub_id)                                                     
    }                                                                           
    subjects_val.append(tio.Subject(**params))                                  
                                                                                
                                                                                
patch_size = (320, 320, 1)                                                      
batch_size = 10                                                                 
                                                                                
sd = tio.SubjectsDataset(subjects)                                              
                                                                                
sampler = tio.data.UniformSampler(patch_size=patch_size)                        
queue = tio.Queue(sd, max_length=40, shuffle_patches=True,                      
        samples_per_volume=-1,                                                  
        sampler=sampler, num_workers=8, shuffle_subjects=True)                  
                                                                                
tr_loader = DataLoader(queue, batch_size=batch_size, shuffle=False,             
        pin_memory=False, num_workers=0)                                        
val_loader = DataLoader(subjects_val, batch_size=batch_size, shuffle=False,        
        pin_memory=False, num_workers=0)                                        
                                                                                
# One epoch of training                                                         
for i, patch in enumerate(tr_loader):                                           
    print(patch["info"])                                                        
                                                                                
    if i % 3 == 0:                                                              
        # Validation                                                            
        for j, _ in enumerate(val_loader):  
            pass
```

### Actual outcome

The code outputs:
['4', '1', '9', '7', '4', '9', '9', '2', '2', '4']
['1', '2', '1', '2', '7', '1', '9', '7', '9', '2']
['7', '4', '1', '1', '2', '2', '1', '7', '4', '7']
['4', '4', '1', '9', '2', '9', '7', '4', '7', '9']
['8', '5', '10', '8', '8', '5', '3', '5', '8', '3']
['6', '3', '3', '3', '10', '6', '10', '5', '10', '10']
['6', '3', '8', '8', '3', '5', '10', '6', '3', '8']
['10', '10', '5', '6', '5', '6', '8', '5', '6', '6']

If we replace "if i % 3 == 0:" with "if False:" (i.e., we don't run the validation loop in the validation set)
['4', '1', '9', '7', '4', '9', '9', '2', '2', '4']
['1', '2', '1', '2', '7', '1', '9', '7', '9', '2']
['7', '4', '1', '1', '2', '2', '1', '7', '4', '7']
['4', '4', '1', '9', '2', '9', '7', '4', '7', '9']
**['8', '6', '3', '8', '5', '6', '10', '10', '5', '10']
['10', '5', '10', '8', '10', '3', '6', '3', '8', '3']
['5', '8', '6', '10', '5', '5', '3', '8', '5', '8']
['10', '3', '8', '5', '6', '3', '6', '6', '6', '3']**

### Expected outcome

The outcome (i.e., the order of the patches in the training set) should be the same regardless of whether we run a validation loop on a separate validation set. This mismatch can affect reproducibility.

### System info

```shell
Platform:   Linux-4.15.0-142-generic-x86_64-with-glibc2.17
TorchIO:    0.18.84
PyTorch:    1.8.1+cu101
SimpleITK:  2.2.0 (ITK 5.3)
NumPy:      1.23.3
Python:     3.8.9 (default, Apr  3 2021, 01:02:10) 
[GCC 5.4.0 20160609]
```

**Comment by jmlipman** [comment]
An even more minimal example: Remove the validation part in the loop, leave only this:
                                            
```
for i, patch in enumerate(tr_loader):                                           
    print(patch["info"])
```

This gives you the data in certain order. Now, add the following line above the for loop:

`_ = enumerate(val_loader)`

That changes the output. And if you add again the same line, it changes again the output, i.e.:

```
_ = enumerate(val_loader)
_ = enumerate(val_loader)
for i, patch in enumerate(tr_loader):                                           
    print(patch["info"])
```

Interestingly, it makes no difference in the output whether you call val_loader or tr_loader in the lines above the for loop.

**Comment by jmlipman** [comment]
Sorry, it turns out that this is not a problem of Torchio.

The problem is because _BaseDataLoaderIter is called when we do iter(dataloader), and the __init__ of this class calls  Tensor.random_() to get a seed. So, when we arrive at the "enumerate(tr_loader)", the "random_" function will be called again, and the seed will be different, leading to a different order.

**Comment by romainVala** [comment]
hi, @jmlipman 
even though it is not a torchio issue, I am curious to know if  you find a solution ?

**Comment by jmlipman** [comment]
I don't think there is any good solution to this. It annoys me a bit because it means that you cannot continue the training exactly where it stops, and, in time-limited servers, it can stop anywhere in the training loop.

It also has interesting implications too: even if you always use the same seed, if you stop the training and continue later, the training samples will be in a different order than if you didn't stop the training. However, although you cannot fully continue the training exactly where it was, it probably doesn't affect the overall performance that much.

One potential solution though, would be to call Tensor.random_() as many times as you called "enumerate(tr_loader)" and "enumerate(val_loader)". This could be done easily based on the epoch/iteration and training set size. But, if you aim to build a complex framework, you may have other calls to Tensor.random_() somewhere else in the code, and you would need to account for those too.
